tap-version-14-specification.md

```diff
@@ -0,0 +1,659 @@
+---
+layout: default
+title: TAP 14 specification
+---
+
+# STATUS: DRAFT
+
+This is a draft document pending further editing and ratification by no
+fewer than 3 widely used TAP implementations in 3 different language
+communities.
+
+This section will likely be removed prior to ratification.
+
+## GOALS OF THIS SPECIFICATION
+
+* Document the observed behavior of widely used TAP implementations.
+* Add no features that are not already in wide usage across multiple
+  implementations.
+* Explicitly allow what is already allowed, deny what is already denied.
+* Provide updated and clearer guidance for new TAP implementations.
+
+# NAME
+
+TAP14 - The Test Anything Protocol v14
+
+## SYNOPSIS
+
+TAP, the Test Anything Protocol, is a simple text-based interface between
+testing modules a test harness. TAP started life as part of the test
+harness for Perl but now has implementations in C/C++, Python, PHP, Perl,
+JavaScript, and probably others by the time you read this.  This document
+describes version 14 of TAP. Go to TAP to read about previous versions.
+
+The key words _must_, _must not_, _required_, _shall_, _shall not_,
+_should_, _should not_, _recommended_, _may_, and _optional_ in this
+document are to be interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+## TAP14 FORMAT
+
+TAP14's general grammar is:
+
+```ebnf
+TAPDocument := Version Plan Body | Version Body Plan
+Version     := "TAP version 14\n"
+Plan        := (Number) ".." (Number) (" # " Reason)? "\n"
+Body        := (TestPoint | BailOut | Pragma | Comment | Anything | Empty)*
```

Comment:
I think technically (as the draft stands right now) in `ok 1 - test # comment`, the `# comment` is an unrecognized directive.  I don't love that.

But that brings up an interesting point to investigate, how to have `#` in a test point name?  This gets weird:

```tap
$ node -e 'require("tap").pass("ok # commenty")'
TAP version 13
ok 1 - ok # commenty
1..1
# time=1.489ms
```

Node's `tap-parser` will interpret any unrecognized directive as part of the description (like you suggest)

```
$ node -e 'require("tap").pass("ok # commenty")' | npx tap-parser
[
  [ 'version', 13 ],
  [
    'assert',
    Result { ok: true, id: 1, name: 'ok # commenty', fullname: '' }
  ],
  ...
```

But then making that test TODO breaks things a bit...

```
$ node -e 'require("tap").pass("ok # commenty", { todo: true })'
TAP version 13
ok 1 - ok # commenty # TODO
1..1
# time=1.48ms
```

```
$ node -e 'require("tap").pass("ok # commenty", { todo: true })' | npx tap-parser
[
  [ 'version', 13 ],
  [
    'assert',
    Result {
      ok: true,
      id: 1,
      name: 'ok # commenty # TODO',
      fullname: ''
    }
```

So, it seems like some options are:

- Specify a way to escape `#` chars, so they're not parsed as directives, and say that producers _should_ escape any `#` characters in test descriptions.  (This seems like a good idea regardless, but I'm not sure if there's consensus on it.  Node-tap will not parse directives that use `\` to escape the `#`.)
- Allow multiple directives (then how can you have # in the `TODO` description?) I don't really want to do this, it gets weird fast 😓
- Do not allow `#` in the test point description (this seems especially ruthless 😬)
- Specify that everything after the first `#` is _the_ directive, and tell harnesses to include unrecognized directives in the description (then this breaks `ok 1 - name # comment # TODO`)
- Specify that everything after the _last_ `#` is _the_ directive (then this breaks `ok 1 - name # TODO # reason`)
- Specify that harnesses should search the string for the first known directive, and treat it as the only directive for that test point, and everything else up to that point is part of the description.  (So `ok 1 - desc # todo # skip` is a todo test with the reason set to `"# skip"`, and `ok 1 - desc # skip # todo` is a skip test with the reason set as `# todo`).
- Just document the edge case and leave it up to harness/producer implementations 🤷  (I'm fine with this, it just kicks the can down the road for implementations to come to consensus for TAP15.)

Reply:
I remember seeing test points with comments somewhere some long time ago, probably in some Perl examples/projects. So I implemented that in tap4j, but it was part of the issues I had when trying to write a formal grammar (ended up writing a horrendous regular expression parser).

I think I'd lean towards either treating it as an ignored directive attached to a test point, or as a comment attached to a test point. It would be easier to avoid hashes within hashes, but in `tap4j` that could happen as that was used in Java, where you can get a method like in a comment like `not ok 1 # SKIP see comment in org.example.Class#methodName(int, String)` (text like this definitely appears in the YAMLish, but users could add it to directives/comments too I guess.)

Reply:
In my reading of the TAP13 spec, the language used was "*starts with* # TODO". In tappy, I took this to mean:

* The first `#` is the start of a directive.
* There can be only one recognized directive per test point.
* Only TODO and SKIP are recognized (with the appropriate case sensitivity and permission of characters after "SKIP" like "Skipped" applied)

The language in the spec calls the stuff after as "explanation" or "reason" depending on the context, so I normalized that part as a "description."

Given all that, I think I would lean toward using the second to last bullet to recognize only a single directive and treat everything after it as the description. Things that don't look like a TODO or SKIP would be an unknown directive (or ignored outright).

Reply:
> Specify a way to escape # chars, so they're not parsed as directives, and say that producers should escape any # characters in test descriptions. (This seems like a good idea regardless, but I'm not sure if there's consensus on it. Node-tap will not parse directives that use \ to escape the #.)

Test::Harness allows for `#`(or `\`) to be escaped with `\`, and has done so since at least 2007

Reply:
Promoting this to an issue: https://github.com/TestAnything/Specification/issues/26